### PR TITLE
Fix rare race condition related to Memory allocation failure

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix extremely rare race condition which could lead to segmentation faults in SELECT queries or endless `Table is readonly` and merges hungs.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit c3dcb81a626b82d711ef230aa9f7d238724d8f91 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56303/c3dcb81a626b82d711ef230aa9f7d238724d8f91/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6746852624">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56303/c3dcb81a626b82d711ef230aa9f7d238724d8f91/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Common/FailPoint.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@ static struct InitFiu
     REGULAR(cluster_discovery_faults) \
     REGULAR(check_table_query_delay_for_part) \
     REGULAR(dummy_failpoint) \
+    REGULAR(prefeteched_reader_pool_failpoint) \
```
Comment (chhetripradeep) [inline src/Common/FailPoint.cpp:42]: Typo: `s/prefeteched/prefetched`

Comment (alesapin) [inline src/Common/FailPoint.cpp:42]:
```suggestion
    REGULAR(prefetched_reader_pool_failpoint) \
```

File: src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp [hunk 2]
```diff
@@ -28,6 +30,11 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
+namespace FailPoints
+{
+    extern const char prefeteched_reader_pool_failpoint[];
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp:35]:
```suggestion
    extern const char prefetched_reader_pool_failpoint[];
```

File: src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp [hunk 3]
```diff
@@ -37,27 +44,51 @@ bool MergeTreePrefetchedReadPool::TaskHolder::operator<(const TaskHolder & other
...
+            prefetch_futures.push_back(pool_.createPrefetchedFuture(reader.get(), priority_));
+
+        fiu_do_on(FailPoints::prefetched_reader_pool_failpoint,
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Failpoint for prefetched reader enabled");
+        });
+    }
+    catch (...) /// in case of memory exceptions we have to wait
+    {
+        for (auto & prefetch_future : prefetch_futures)
```
Comment (antonio2368) [inline src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp:76]: So we create a future with 
```
pool_.createPrefetchedFuture(readers.main.get(), priority_))
```
and then push it to the vector
what if we get OOM when allocating extra space inside vector?
we won't wait for the created future AFAIS

Comment (antonio2368) [reply src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp:76]: I think simple `reserve` is enough to avoid it

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**